runtime/nvqir/cutensornet/tensornet_state.cpp

```diff
@@ -0,0 +1,371 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "tensornet_state.h"
+#include <cassert>
+
+namespace nvqir {
+
+TensorNetState::TensorNetState(std::size_t numQubits,
+                               cutensornetHandle_t handle)
+    : m_numQubits(numQubits), m_cutnHandle(handle) {
+  const std::vector<int64_t> qubitDims(m_numQubits, 2);
+  HANDLE_CUTN_ERROR(cutensornetCreateState(
+      m_cutnHandle, CUTENSORNET_STATE_PURITY_PURE, m_numQubits,
+      qubitDims.data(), CUDA_C_64F, &m_quantumState));
+}
+
+void TensorNetState::applyGate(const std::vector<int32_t> &qubitIds,
+                               void *gateDeviceMem, bool adjoint) {
+
+  int64_t id;
+  HANDLE_CUTN_ERROR(cutensornetStateApplyTensor(
+      m_cutnHandle, m_quantumState, qubitIds.size(), qubitIds.data(),
+      gateDeviceMem, nullptr, /*immutable*/ 1,
+      /*adjoint*/ static_cast<int32_t>(adjoint), /*unitary*/ 1, &id));
```

Comment:
For reasons mentioned before, I would recommend initializing `id` this to 0. (I believe that `cutensornetStateApplyTensor()` will not allow a `nullptr`, so that is not an option here.)

Reply:
Good point. Fixed as suggested.